File: robot-factory/typescript/src/PartSupplier.ts
import type { PartOption } from './PartOption.js';
import type { PartQuote } from './PartQuote.js';
import type { PartType } from './PartType.js';
import type { PurchasedPart } from './PurchasedPart.js';

export interface PartSupplier {
  readonly name: string;
  getQuote(type: PartType, option: PartOption): PartQuote | null;
  purchase(type: PartType, option: PartOption): PurchasedPart;
}

File: robot-factory/python/src/robot_factory/part_supplier.py
from __future__ import annotations

from typing import Protocol

from .part_option import PartOption
from .part_quote import PartQuote
from .part_type import PartType
from .purchased_part import PurchasedPart


class PartSupplier(Protocol):
    @property
    def name(self) -> str: ...

    def get_quote(self, part_type: PartType, option: PartOption) -> PartQuote | None: ...

    def purchase(self, part_type: PartType, option: PartOption) -> PurchasedPart: ...
